scripts/builtin/steplm.dml

```diff
@@ -0,0 +1,320 @@
+# hadoop jar SystemML.jar -f StepLinearRegDS.dml -nvargs X=INPUT_DIR/X Y=INPUT_DIR/Y B=OUTPUT_DIR/betas
+#     O=OUTPUT_DIR/stats S=OUTPUT_DIR/selected icpt=2 thr=0.01 fmt=csv write_beta=TRUE
+
+
+m_steplm = function(Matrix[Double] X, Matrix[Double] y, Integer icpt = 0, Double reg = 1e-7, Double tol = 1e-7, Integer maxi = 0, Boolean verbose = TRUE)
+  return() {
+
+
+  # currently only the forward selection strategy in supported: start from one feature and iteratively add 
+  # features until AIC improves
+  dir = "forward";
+
+  #fmt = ifdef($fmt, "text");
+  #icpt = intercept_status;
+  thr = ifdef($thr, 0.001);
+ 
+
+  print("BEGIN STEPWISE LINEAR REGRESSION SCRIPT");
+  print("Reading X and Y...");
+  X_orig = X;
+
+  n = nrow(X_orig);
+  m_orig = ncol(X_orig);
+
+  # BEGIN STEPWISE LINEAR REGRESSION
+
+  if (dir == "forward") {
+    continue = TRUE;
+    columns_fixed = matrix(0, rows = 1, cols = m_orig);
+    columns_fixed_ordered = matrix(0, rows = 1, cols = 1);
+
+    # X_global stores the best model found at each step
+    X_global = matrix(0, rows = n, cols = 1);
+
+    if (icpt == 1 | icpt == 2) {
+      beta = mean(y);
+      AIC_best = 2 + n * log(sum((beta - y) ^ 2) / n);
+    } else {
+      beta = 0.0;
+      AIC_best = n * log(sum(y ^ 2) / n);
+    }
+
+    AICs = matrix(AIC_best, rows = 1, cols = m_orig);
+    print("Best AIC without any features: " + AIC_best);
+
+    boa_ncol = ncol(X_orig);
+    if (icpt != 0) {
+      boa_ncol = boa_ncol + 1
+    }
+
+    beta_out_all = matrix(0, rows = boa_ncol, cols = m_orig * 1);
+
+    y_ncol = 1;
+    column_best = 0;
+    # First pass to examine single features
+    for (i in 1:m_orig, check = 0) {
+      columns_fixed_ordered_1 = matrix(i, rows = 1, cols = 1);
+      [AIC_1, beta_out_i] = linear_regression(X_orig[, i], y, icpt);
+
+      print(" mibeta "  + toString(beta_out_i))
+      AICs[1, i] = AIC_1;
+      AIC_cur = as.scalar(AICs[1, i]);
+      if ((AIC_cur < AIC_best) & ((AIC_best - AIC_cur) > abs(thr * AIC_best))) {
+        column_best = i;
+        AIC_best = as.scalar(AICs[1, i]);
+      }
+
+      beta_out_all[1:nrow(beta_out_i), (i - 1) * y_ncol + 1:i * y_ncol] = beta_out_i[, 1:1];
+
+    }
+
+    # beta best so far
+    beta_best = beta_out_all[, (column_best - 1) * y_ncol + 1:column_best * y_ncol];
+
+    if (column_best == 0) {
+      print("AIC of an empty model is " + AIC_best + " and adding no feature achieves more than " +
+           (thr * 100) + "% decrease in AIC!");
+      Selected = matrix(0, rows = 1, cols = 1);
+      if (icpt == 0) {
+        B = matrix(beta, rows = m_orig, cols = 1);
+      } else {
+        B_tmp = matrix(0, rows = m_orig + 1, cols = 1);
+        B_tmp[m_orig + 1,] = beta;
+        B = B_tmp;
+      }
+
+      beta_out = B;
+
+      print("estoy en el stop");
+      print(toString(beta_out) + " mi beta out");
+      print("Selected " + toString(Selected));
+
+      write(Selected, $4);
+      write(beta_out, $3);
+   
```

Comment:
Please make all write return arguments - this function should be (except for prints) side effect free.